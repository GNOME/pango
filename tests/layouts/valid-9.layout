{
  "context" : {
    "font" : "serif 12",
    "language" : "en-us",
    "base-gravity" : "south",
    "gravity-hint" : "natural",
    "base-dir" : "weak-ltr",
    "round-glyph-positions" : true,
    "transform" : [
      1,
      0,
      0,
      1,
      0,
      0
    ]
  },
  "comment" : "Letterspacing with scripts",
  "text" : "Hello שלום Γειά σας",
  "attributes" : [
    {
      "end" : 30,
      "type" : "letter-spacing",
      "value" : 8888
    }
  ],
  "font" : "Cantarell 11",
  "width" : 102400,
  "output" : {
    "is-wrapped" : true,
    "is-ellipsized" : false,
    "unknown-glyphs" : 0,
    "width" : 71392,
    "height" : 77824,
    "log-attrs" : [
      {
        "char-break" : true,
        "cursor-position" : true,
        "word-start" : true,
        "sentence-boundary" : true,
        "sentence-start" : true,
        "backspace-deletes-character" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "word-end" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-start" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "backspace-deletes-character" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "backspace-deletes-character" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "backspace-deletes-character" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "word-end" : true,
        "backspace-deletes-character" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-start" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "word-end" : true,
        "expandable-space" : true,
        "word-boundary" : true
      },
      {
        "line-break" : true,
        "char-break" : true,
        "cursor-position" : true,
        "word-start" : true,
        "word-boundary" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "char-break" : true,
        "cursor-position" : true,
        "break-inserts-hyphen" : true
      },
      {
        "line-break" : true,
        "mandatory-break" : true,
        "char-break" : true,
        "white" : true,
        "cursor-position" : true,
        "word-end" : true,
        "sentence-boundary" : true,
        "sentence-end" : true,
        "word-boundary" : true
      }
    ],
    "lines" : [
      {
        "start-index" : 0,
        "length" : 6,
        "paragraph-start" : true,
        "direction" : "ltr",
        "runs" : [
          {
            "offset" : 0,
            "length" : 6,
            "text" : "Hello ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "en-us",
            "script" : "latin",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "extra-attributes" : [
              {
                "end" : 30,
                "type" : "letter-spacing",
                "value" : 8888
              }
            ],
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 76,
                "width" : 15708,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 287,
                "width" : 17080,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 1
              },
              {
                "glyph" : 349,
                "width" : 12984,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 2
              },
              {
                "glyph" : 349,
                "width" : 12984,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 3
              },
              {
                "glyph" : 370,
                "width" : 12636,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 4
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 5
              }
            ]
          }
        ]
      },
      {
        "start-index" : 6,
        "length" : 9,
        "paragraph-start" : false,
        "direction" : "ltr",
        "runs" : [
          {
            "offset" : 6,
            "length" : 8,
            "text" : "שלום",
            "bidi-level" : 1,
            "gravity" : "south",
            "language" : "he",
            "script" : "hebrew",
            "font" : {
              "description" : "DejaVu Sans 11",
              "checksum" : "69ccd07023a72ceb27a5e5c22f728627353b60a198170f5e58dd7014221abf01",
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "extra-attributes" : [
              {
                "end" : 30,
                "type" : "letter-spacing",
                "value" : 8888
              }
            ],
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 1332,
                "width" : 14684,
                "is-cluster-start" : true,
                "log-cluster" : 6
              },
              {
                "glyph" : 1324,
                "width" : 12984,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 4
              },
              {
                "glyph" : 1331,
                "width" : 17080,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 2
              },
              {
                "glyph" : 1344,
                "width" : 19128,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 0
              }
            ]
          },
          {
            "offset" : 14,
            "length" : 1,
            "text" : " ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "he",
            "script" : "hebrew",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "extra-attributes" : [
              {
                "end" : 30,
                "type" : "letter-spacing",
                "value" : 8888
              }
            ],
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 268435455,
                "width" : 0,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 0
              }
            ]
          }
        ]
      },
      {
        "start-index" : 15,
        "length" : 9,
        "paragraph-start" : false,
        "direction" : "ltr",
        "runs" : [
          {
            "offset" : 15,
            "length" : 9,
            "text" : "Γειά ",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "el",
            "script" : "greek",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "extra-attributes" : [
              {
                "end" : 30,
                "type" : "letter-spacing",
                "value" : 8888
              }
            ],
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 677,
                "width" : 12636,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 808,
                "width" : 16056,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 2
              },
              {
                "glyph" : 812,
                "width" : 12984,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 4
              },
              {
                "glyph" : 837,
                "width" : 13660,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 6
              },
              {
                "glyph" : 268435455,
                "width" : 0,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 8
              }
            ]
          }
        ]
      },
      {
        "start-index" : 24,
        "length" : 6,
        "paragraph-start" : false,
        "direction" : "ltr",
        "runs" : [
          {
            "offset" : 24,
            "length" : 6,
            "text" : "σας",
            "bidi-level" : 0,
            "gravity" : "south",
            "language" : "el",
            "script" : "greek",
            "font" : {
              "description" : "Cantarell 11",
              "checksum" : "5bcb6ee14ee9d210b2e91d643de1fe456e9d1aea770983fdb05951545efebbe2",
              "variations" : {
                "wght" : 0
              },
              "matrix" : [
                1,
                -0,
                -0,
                1,
                0,
                0
              ]
            },
            "flags" : 0,
            "extra-attributes" : [
              {
                "end" : 30,
                "type" : "letter-spacing",
                "value" : 8888
              }
            ],
            "y-offset" : 0,
            "start-x-offset" : 0,
            "end-x-offset" : 0,
            "glyphs" : [
              {
                "glyph" : 822,
                "width" : 13660,
                "is-cluster-start" : true,
                "log-cluster" : 0
              },
              {
                "glyph" : 804,
                "width" : 18104,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 2
              },
              {
                "glyph" : 821,
                "width" : 11612,
                "x-offset" : 4444,
                "is-cluster-start" : true,
                "log-cluster" : 4
              }
            ]
          }
        ]
      }
    ]
  }
}
