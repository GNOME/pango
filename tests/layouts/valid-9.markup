width=100
<span letter_spacing="8888">Hello שלום Γειά σας</span>
