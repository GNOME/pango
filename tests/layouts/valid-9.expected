Hello שלום Γειά σας

--- parameters

wrapped: 1
ellipsized: 0
lines: 5
width: 102400

--- attributes

range 0 30
[0,30]letter-spacing=8888
range 30 2147483647

--- directions

0 0 0 0 0 0 1 1 1 1 0 0 0 0 0 0 0 0 0 0 

--- cursor positions

0(0) 1(0) 2(0) 3(0) 4(0) 5(0) 5(1) 12(0) 10(0) 8(0) 14(0) 14(1) 17(0) 19(0) 21(0) 23(0) 23(1) 26(0) 28(0) 28(1) 31(0) 

--- lines

i=1, index=0, paragraph-start=1, dir=ltr 'Hello שלו'
i=2, index=12, paragraph-start=0, dir=ltr 'ם '
i=3, index=15, paragraph-start=0, dir=ltr 'Γειά '
i=4, index=24, paragraph-start=0, dir=ltr 'σας
'
i=5, index=31, paragraph-start=1, dir=ltr ''

--- runs

i=1, index=0, chars=6, level=0, gravity=south, flags=0, font=OMITTED, script=latin, language=en-us, 'Hello '
[0,30]letter-spacing=8888
i=2, index=6, no run, line end
i=3, index=12, chars=4, level=1, gravity=south, flags=0, font=OMITTED, script=hebrew, language=he, 'שלום'
[0,30]letter-spacing=8888
i=4, index=14, chars=1, level=0, gravity=south, flags=0, font=OMITTED, script=hebrew, language=he, ' '
[0,30]letter-spacing=8888
i=5, index=15, no run, line end
i=6, index=15, chars=5, level=0, gravity=south, flags=0, font=OMITTED, script=greek, language=el, 'Γειά '
[0,30]letter-spacing=8888
i=7, index=24, no run, line end
i=8, index=24, chars=3, level=0, gravity=south, flags=0, font=OMITTED, script=greek, language=el, 'σας'
[0,30]letter-spacing=8888
i=9, index=30, no run, line end
i=10, index=31, no run, line end
